REGRESSION (286055@main): Relative-positioned input elements in scroll areas don't render outlines

smfr opened the pull request:
#### d183e4fe4bcd3f33ca348129099c1d60e6361965
<pre>
REGRESSION (286055@main): Relative-positioned input elements in scroll areas don&apos;t render outlines
<a href="https://bugs.webkit.org/show_bug.cgi?id=286002">https://bugs.webkit.org/show_bug.cgi?id=286002</a>
<a href="https://rdar.apple.com/142995142">rdar://142995142</a>

Reviewed by Alan Baradlay.

286055@main changed the paint phases in overflow scroll in such a way that we never painted
outlines in RenderLayers where the `PaintingOverflowContents` flag was true. This flag is set
for a layer with overflow scrolling, and all its descendent layers.

The `shouldPaintOutline()` lambda returns false if `PaintingOverflowContents` is set, with
the intention of avoid the double-painting of outlines on both the outer primary layer
of a scroller, and its inner scrolled contents layer. However, we need to ensure that
we do paint outlines on child RenderLayers.

To fix this, introduce a `PaintingOverflowContentsRoot` flag that is set only when painting
from the scrolled contents GraphicsLayer, and then cleared when RenderLayer painting traverses
to descendants.

* LayoutTests/compositing/overflow/outline-paint-phases-expected.html: Added.
* LayoutTests/compositing/overflow/outline-paint-phases.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::paintLayerContentsAndReflection):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintOutlineForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintFlagsForLayer const):

Canonical link: <a href="https://commits.webkit.org/289154@main">https://commits.webkit.org/289154@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e57ae7efa4e56cc14a04b8ebc964f6df163dd5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85340 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39771 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90468 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36382 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87429 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5164 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66344 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24153 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88386 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77516 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46626 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3818 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31777 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35450 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74539 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32615 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91977 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12687 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9276 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74946 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12916 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73353 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74066 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18349 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18418 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16851 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4708 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12630 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18095 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12460 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14211 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas reviewed (approved):
It works! Thanks for the patch

smfr committed:
REGRESSION (286055@main): Relative-positioned input elements in scroll areas don't render outlines
https://bugs.webkit.org/show_bug.cgi?id=286002
rdar://142995142

Reviewed by Alan Baradlay.

286055@main changed the paint phases in overflow scroll in such a way that we never painted
outlines in RenderLayers where the `PaintingOverflowContents` flag was true. This flag is set
for a layer with overflow scrolling, and all its descendent layers.

The `shouldPaintOutline()` lambda returns false if `PaintingOverflowContents` is set, with
the intention of avoid the double-painting of outlines on both the outer primary layer
of a scroller, and its inner scrolled contents layer. However, we need to ensure that
we do paint outlines on child RenderLayers.

To fix this, introduce a `PaintingOverflowContentsRoot` flag that is set only when painting
from the scrolled contents GraphicsLayer, and then cleared when RenderLayer painting traverses
to descendants.

* LayoutTests/compositing/overflow/outline-paint-phases-expected.html: Added.
* LayoutTests/compositing/overflow/outline-paint-phases.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::paintLayerContentsAndReflection):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::paintOutlineForFragments):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintFlagsForLayer const):

Canonical link: https://commits.webkit.org/289154@main
